### PR TITLE
[Fix] Prevent passing in too many public inputs to Varuna

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

```
cd synthesizer/snark
cargo test test_varuna_verify_public_input_size --lib
```

### CODE REVIEW

**Comment by vicsn** [comment]
Some notes about the fixes:
- handling errors in nested closures is very annoying, so I replaced some with simple for-loops
~- we don't use `ensure!` anywhere yet in `varuna.rs`, so followed the existing setup when throwing a `SNARKError::PublicInputSizeMismatch`~

File: algorithms/src/snark/varuna/tests.rs [hunk 1]
```diff
@@ -76,7 +76,7 @@ mod varuna {
                         assert!($snark_inst::verify(universal_verifier, &fs_parameters, &index_vk, public_inputs, &proof).unwrap());
                         println!("Called verifier");
                         eprintln!("\nShould not verify (i.e. verifier messages should print below):");
-                        assert!(!$snark_inst::verify(universal_verifier, &fs_parameters, &index_vk, [random, random], &proof).unwrap());
+                        assert!(!$snark_inst::verify(universal_verifier, &fs_parameters, &index_vk, [random], &proof).unwrap());
```
Comment (howardwu) [inline algorithms/src/snark/varuna/tests.rs]: Why was this changed?

File: algorithms/src/snark/varuna/varuna.rs [hunk 2]
```diff
@@ -670,7 +670,13 @@ where
 
             let input_fields = public_inputs_i
                 .iter()
-                .map(|input| input.borrow().to_field_elements())
+                .map(|input| {
+                    let input = input.borrow().to_field_elements()?;
+                    if input.len() > input_domain.size() - 1 {
```
Comment (ljedrz) [inline algorithms/src/snark/varuna/varuna.rs]: is input domain size guaranteed to be non-zero?

File: algorithms/src/snark/varuna/ahp/ahp.rs [hunk 3]
```diff
@@ -180,23 +180,20 @@ impl<F: PrimeField, SM: SNARKMode> AHPForR1CS<F, SM> {
...
+            let input_domain = circuit_state.input_domain;
+            let public_inputs_i = public_inputs[circuit_id]
+                .iter()
+                .map(|p| {
+                    let public_input = prover::ConstraintSystem::format_public_input(p);
+                    Self::formatted_public_input_is_admissible(&public_input)?;
+                    Ok::<_, AHPError>(public_input)
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            assert_eq!(public_inputs_i[0].len(), input_domain.size());
```
Comment (ljedrz) [inline algorithms/src/snark/varuna/ahp/ahp.rs]: shouldn't this remain `ensure`, since we already have error handling available here?

**Review by ljedrz — APPROVED** [review]
LGTM with 2 questions addressed :+1:

File: algorithms/src/snark/varuna/tests.rs (hunk 1 above)
Comment (vicsn) [reply algorithms/src/snark/varuna/tests.rs]: The new test and code introduced in this PR succesfully prevents too many inputs from being passed in. We used to pass in two public inputs here, even though the circuit only expects one public input.

For extra context: this is a negative test, so we're passing a random public input and expect it to fail.

File: algorithms/src/snark/varuna/ahp/ahp.rs (hunk 3 above)
Comment (vicsn) [reply algorithms/src/snark/varuna/ahp/ahp.rs]: Good catch: [Use ensure instead of assert_eq](https://github.com/AleoHQ/snarkVM/pull/2094/commits/e0ea4593decc3f608e015c283222676210f63bad)

File: algorithms/src/snark/varuna/varuna.rs (hunk 2 above)
Comment (vicsn) [reply algorithms/src/snark/varuna/varuna.rs]: This let me down an interesting rabbit hole. Yes it is guaranteed to be non-zero because we're always passing in a public input. And the - 1 is no longer needed in this somewhat related PR: https://github.com/AleoHQ/snarkVM/pull/2268

**Comment by howardwu** [comment]
@vicsn is this PR still relevant?

**Comment by vicsn** [comment]
> @vicsn is this PR still relevant?

@howardwu yes it is, made the PR title more descriptive.

File: algorithms/src/snark/varuna/varuna.rs (hunk 2 above)
Comment (howardwu) [reply algorithms/src/snark/varuna/varuna.rs]: So is the -1 still needed here?

Comment (vicsn) [reply algorithms/src/snark/varuna/varuna.rs]: Nope, merged in the `mainnet` branch and resolved the conflict.

**Comment by howardwu** [comment]
LGTM
